Sources/DataStructures/SegmentTree.swift

```diff
@@ -0,0 +1,130 @@
+//
+//  SegmentTree.swift
+//  DataStructures
+//
+//  Created by James Bean on 12/4/17.
+//
+
+import Combinatorics
+import Destructure
+
+public struct Interval <Metric: SignedNumeric & Comparable> {
+
+    enum Bound {
+        case start
+        case end
+    }
+
+    let start: Metric
+    let end: Metric
+
+    public init(_ start: Metric, _ end: Metric) {
+        precondition(end >= start)
+        self.start = start
+        self.end = end
+    }
+}
+
+extension Interval: Equatable {
+    public static func == (lhs: Interval, rhs: Interval) -> Bool {
+        return lhs.start == rhs.start && lhs.end == rhs.end
+    }
+}
+
+enum SegmentTree <Metric: SignedNumeric & Comparable> {
+
+    public var interval: Interval<Metric> {
+        switch self {
+        case .leaf(let interval):
+            return interval
+        case .branch(_, let interval, _):
+            return interval
+        }
+    }
+
+    case leaf(Interval<Metric>)
+    indirect case branch(SegmentTree, Interval<Metric>, SegmentTree)
+
+    init(_ a: SegmentTree, _ b: SegmentTree) {
+        precondition(a.interval.end == b.interval.start)
+        self = .branch(a, Interval(a.interval.start, b.interval.end), b)
+    }
+
+    init(_ interval: Interval<Metric>) {
+        self = .leaf(interval)
+    }
+
+    init(offsets: [Metric]) {
+
+        func clump(_ trees: [SegmentTree]) -> ([(SegmentTree, SegmentTree)], SegmentTree?) {
+            precondition(!trees.isEmpty)
+            let pairs = stride(from: trees.startIndex, to: trees.endIndex - 1, by: 2).map { index in
+                return (trees[index], trees[index + 1])
+            }
+            let last = trees.count % 2 != 0 ? trees.last! : nil
+            return (pairs,last)
+        }
+
+        func join(_ trees: [SegmentTree]) -> SegmentTree {
+            guard trees.count > 1 else { return trees.first! }
+            let (pairs,last) = clump(trees)
+            let left = join(pairs.map(SegmentTree.init))
+            if let last = last {
+                return SegmentTree(left,last)
+            } else {
+                return left
+            }
+        }
+
+        switch offsets.count {
+        case 0,1:
+            fatalError("Not enough points to create an IntervalTree")
+        default:
+            self = join(offsets.pairs.map(Interval.init).map(SegmentTree.init))
+        }
+    }
+
+    subscript(interval: Interval<Metric>) -> SegmentTree {
```

Comment:
what's the intended usage for subscript? seems like there are so many ways to implement it. it could even return `[SegmentTree]`. I think I would expect it to be either (1) an (ordered) list of the leaves whose given intervals fit inside `interval` (which could have a pretty nice recursive design), or (2) the largest `SegmentTree` that fits inside `interval`, with a precondition/error throw if none exists (which could also have a nice recursion).